File: tools/measure_viewer.py
#!/usr/bin/env python3
"""Viewer-side cost of a fux binary: the real `fux` viewer on a pty against an isolated server.

Usage: tools/measure_viewer.py PATH/TO/fux [--keystrokes 200] [--rows 24] [--columns 80]

Reports the viewer process's CPU seconds and the bytes it wrote to its terminal for a keystroke
loop (one echoed byte at a time) and for a 20 000-line output burst, plus the server's CPU for
the same phases. The pty is drained continuously so the viewer never blocks on a full buffer.
Nothing here touches personal sessions.
"""
import argparse
import fcntl
import json
import os
import pty
import re
import select
import signal
import struct
import subprocess
import sys
import tempfile
import termios
import time
from pathlib import Path

ESCAPES = re.compile(rb'\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)|\x1b\[[0-9;?<>=]*[ -/]*[@-~]|\x1b[()][A-Za-z0-9]|\x1b[=>]')


def cpu_seconds(pid):
    out = subprocess.run(["ps", "-o", "cputime=", "-p", str(pid)], capture_output=True, text=True).stdout.strip()
    total = 0.0
    for part in out.replace("-", ":").split(":"):
        total = total * 60 + float(part)
    return total


class Terminal:
    def __init__(self, binary, env, rows, columns):
        self.raw = b""
        self.read_bytes = 0
        self.pid, self.fd = pty.fork()
        if self.pid == 0:
            fcntl.ioctl(0, termios.TIOCSWINSZ, struct.pack("HHHH", rows, columns, 0, 0))
            os.execve(binary, [binary], env)

    def pump(self, seconds):
        end = time.monotonic() + seconds
        while time.monotonic() < end:
            ready, _, _ = select.select([self.fd], [], [], max(0, end - time.monotonic()))
            if not ready:
                continue
            try:
                data = os.read(self.fd, 65536)
            except OSError:
                return False
            if not data:
                return False
            self.raw += data
            self.raw = self.raw[-(1 << 20):]
            self.read_bytes += len(data)
        return True

    def text(self):
        return ESCAPES.sub(b"", self.raw)

    def wait_for(self, predicate, timeout):
        end = time.monotonic() + timeout
        while time.monotonic() < end:
            self.pump(0.02)
            if predicate(self.text()):
                return
        raise TimeoutError("viewer output did not arrive")

    def send(self, data):
        os.write(self.fd, data)

    def close(self):
        try:
            os.kill(self.pid, signal.SIGKILL)
            os.waitpid(self.pid, 0)
        except OSError:
            pass
        os.close(self.fd)


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("binary")
    parser.add_argument("--keystrokes", type=int, default=200)
    parser.add_argument("--rows", type=int, default=24)
    parser.add_argument("--columns", type=int, default=80)
    args = parser.parse_args()
    binary = str(Path(args.binary).resolve())
    with tempfile.TemporaryDirectory(prefix="fux-viewer-", dir="/tmp") as directory:
        root = Path(directory)
        env = os.environ.copy()
        env.update(HOME=directory, XDG_RUNTIME_DIR=directory, XDG_CONFIG_HOME=directory + "/config",
                   XDG_STATE_HOME=directory + "/state", SHELL="/bin/sh", TERM="xterm-256color")
        (root / "config/fux").mkdir(parents=True)
        (root / "config/fux/config.toml").write_text('default-command = { argv = ["/bin/sh", "-c", "printf READY; exec /bin/sh"] }\n')
        server = subprocess.Popen([binary, "serve", "--name", "default"], env=env,
                                  stdout=subprocess.DEVNULL, stderr=subprocess.PIPE)
        terminal = None
        try:
            sock_path = root / "fux/default.attach.sock"
            deadline = time.monotonic() + 10
            while time.monotonic() < deadline and not sock_path.exists():
                time.sleep(0.005)
            terminal = Terminal(binary, env, args.rows, args.columns)
            terminal.wait_for(lambda text: b"READY" in text, 10)
            terminal.pump(0.5)
            # Keystrokes.
            viewer_cpu = cpu_seconds(terminal.pid)
            server_cpu = cpu_seconds(server.pid)
            bytes_before = terminal.read_bytes
            begin = time.monotonic()
            for index in range(args.keystrokes):
                if index and index % 40 == 0:
                    terminal.raw = b""
                    terminal.send(b"\x15")
                    terminal.wait_for(lambda text: b"a" not in text[-200:], 5)
                    terminal.raw = b""
                wanted = index % 40 + 1
                terminal.send(b"a")
                terminal.wait_for(lambda text, wanted=wanted: text.count(b"a") >= wanted, 5)
            keystrokes_s = time.monotonic() - begin
            viewer_cpu_keys = cpu_seconds(terminal.pid) - viewer_cpu
            server_cpu_keys = cpu_seconds(server.pid) - server_cpu
            bytes_keys = terminal.read_bytes - bytes_before
            terminal.send(b"\x15")
            terminal.pump(0.3)
            # Burst.
            terminal.raw = b""
            viewer_cpu = cpu_seconds(terminal.pid)
            server_cpu = cpu_seconds(server.pid)
            bytes_before = terminal.read_bytes
            terminal.send(b"i=0; while [ $i -lt 20000 ]; do echo line$i; i=$((i+1)); done; printf BURST''DONE\\\\n\n")
            begin = time.monotonic()
            terminal.wait_for(lambda text: b"BURSTDONE" in text, 60)
            burst_s = time.monotonic() - begin
            terminal.pump(0.3)
            print(json.dumps({
                "binary": binary,
                "size": f"{args.rows}x{args.columns}",
                "keystrokes": args.keystrokes,
                "keystroke_loop_s": round(keystrokes_s, 3),
                "viewer_cpu_s_per_1000_keystrokes": round(viewer_cpu_keys * 1000 / args.keystrokes, 3),
                "server_cpu_s_per_1000_keystrokes": round(server_cpu_keys * 1000 / args.keystrokes, 3),
                "pty_bytes_per_keystroke": bytes_keys // args.keystrokes,
                "burst_s": round(burst_s, 3),
                "burst_viewer_cpu_s": round(cpu_seconds(terminal.pid) - viewer_cpu, 3),
                "burst_server_cpu_s": round(cpu_seconds(server.pid) - server_cpu, 3),
                "burst_pty_bytes": terminal.read_bytes - bytes_before,
            }, indent=2))
            terminal.send(b"\x01d")
            terminal.pump(1)
        finally:
            if terminal is not None:
                terminal.close()
            server.terminate()
            try:
                server.wait(timeout=10)
            except subprocess.TimeoutExpired:
                server.kill()
                server.wait()


if __name__ == "__main__":
    sys.exit(main())

File: tools/xtask/src/measure_viewer.rs
//! Main's real-viewer PTY benchmark. Text is an escape-stripped transcript, not a screen.
use crate::measure::{cpu, round};
use anyhow::{Context, Result, ensure};
use fux_xtask::support::{local::Root, terminal::Terminal};
use serde_json::{Value, json};
use std::time::{Duration, Instant};

struct Reader {
    terminal: Terminal,
    raw: Vec<u8>,
    bytes: u64,
    escapes: regex::bytes::Regex,
}
impl Reader {
    fn pump(&mut self, duration: Duration) -> Result<()> {
        self.terminal.pump_for(duration, |bytes| {
            self.bytes += bytes.len() as u64;
            self.raw.extend_from_slice(bytes);
            let excess = self.raw.len().saturating_sub(1 << 20);
            self.raw.drain(..excess);
        })
    }
    fn wait_for(&mut self, predicate: impl Fn(&[u8]) -> bool, timeout: Duration) -> Result<()> {
        let deadline = Instant::now() + timeout;
        while Instant::now() < deadline {
            self.pump(
                Duration::from_millis(20).min(deadline.saturating_duration_since(Instant::now())),
            )?;
            if predicate(&self.escapes.replace_all(&self.raw, &b""[..])) {
                return Ok(());
            }
            ensure!(
                self.terminal.child.0.try_wait()?.is_none(),
                "viewer exited before benchmark marker"
            );
        }
        anyhow::bail!("viewer output did not arrive")
    }
}
fn contains(text: &[u8], marker: &[u8]) -> bool {
    text.windows(marker.len()).any(|window| window == marker)
}

pub fn run(args: Vec<String>) -> Result<()> {
    let mut args = args.into_iter();
    let binary = std::fs::canonicalize(
        args.next()
            .context("measure-viewer BINARY [--keystrokes N] [--rows N] [--columns N]")?,
    )?;
    let mut keystrokes: usize = 200;
    let mut rows: u16 = 24;
    let mut columns: u16 = 80;
    while let Some(arg) = args.next() {
        let value = args.next().context("missing option value")?;
        match arg.as_str() {
            "--keystrokes" => keystrokes = value.parse()?,
            "--rows" => rows = value.parse()?,
            "--columns" => columns = value.parse()?,
            _ => anyhow::bail!("unknown measure-viewer option: {arg}"),
        }
    }
    ensure!(
        (1..=10000).contains(&keystrokes) && rows > 0 && columns > 0,
        "invalid viewer benchmark dimensions or keystrokes"
    );
    let root = Root::new(
        "fux-viewer-",
        &[
            "/bin/sh".into(),
            "-c".into(),
            "printf READY; exec /bin/sh".into(),
        ],
    )?;
    let mut server = root.server(&binary)?;
    let result: Result<Value> = (|| {
        let terminal = Terminal::start_with_size(&root, &binary, &[], rows, columns)?;
        let mut reader = Reader {
            terminal,
            raw: Vec::new(),
            bytes: 0,
            escapes: regex::bytes::Regex::new(
                r"\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)|\x1b\[[0-9;?<>=]*[ -/]*[@-~]|\x1b[()][A-Za-z0-9]|\x1b[=>]",
            )?,
        };
        reader.wait_for(|text| contains(text, b"READY"), Duration::from_secs(10))?;
        reader.pump(Duration::from_millis(500))?;
        reader.raw.clear();
        let viewer = reader.terminal.child.0.id();
        let server_pid = server.child.id();
        let viewer_cpu = cpu(viewer)?;
        let server_cpu = cpu(server_pid)?;
        let before = reader.bytes;
        let begin = Instant::now();
        for index in 0..keystrokes {
            if index > 0 && index % 40 == 0 {
                reader.raw.clear();
                reader.terminal.send(&[0x15])?;
                reader.wait_for(
                    |text| !text[text.len().saturating_sub(200)..].contains(&b'a'),
                    Duration::from_secs(5),
                )?;
                reader.raw.clear();
            }
            let wanted = index % 40 + 1;
            reader.terminal.send(b"a")?;
            reader.wait_for(
                |text| text.iter().filter(|&&byte| byte == b'a').count() >= wanted,
                Duration::from_secs(5),
            )?;
        }
        let keys = begin.elapsed().as_secs_f64();
        let viewer_keys = cpu(viewer)? - viewer_cpu;
        let server_keys = cpu(server_pid)? - server_cpu;
        let bytes_keys = reader.bytes - before;
        reader.terminal.send(&[0x15])?;
        reader.pump(Duration::from_millis(300))?;
        reader.raw.clear();
        let viewer_cpu = cpu(viewer)?;
        let server_cpu = cpu(server_pid)?;
        let before = reader.bytes;
        reader.terminal.send(b"i=0; while [ $i -lt 20000 ]; do echo line$i; i=$((i+1)); done; printf BURST''DONE\\\\n\n")?;
        let begin = Instant::now();
        reader.wait_for(|text| contains(text, b"BURSTDONE"), Duration::from_secs(60))?;
        let burst = begin.elapsed().as_secs_f64();
        reader.pump(Duration::from_millis(300))?;
        let result = json!({"binary":binary,"size":format!("{rows}x{columns}"),"keystrokes":keystrokes,
            "keystroke_loop_s":round(keys,3),"viewer_cpu_s_per_1000_keystrokes":round(viewer_keys*1000./keystrokes as f64,3),
            "server_cpu_s_per_1000_keystrokes":round(server_keys*1000./keystrokes as f64,3),"pty_bytes_per_keystroke":bytes_keys/keystrokes as u64,
            "burst_s":round(burst,3),"burst_viewer_cpu_s":round(cpu(viewer)?-viewer_cpu,3),
            "burst_server_cpu_s":round(cpu(server_pid)?-server_cpu,3),"burst_pty_bytes":reader.bytes-before});
        reader.terminal.send(b"\x01d")?;
        reader.pump(Duration::from_secs(1))?;
        reader.terminal.close()?;
        Ok(result)
    })();
    let cleanup = server.finish();
    let result = result?;
    cleanup?;
    println!("{}", serde_json::to_string_pretty(&result)?);
    Ok(())
}
